File: src/audio/mapping.ts
import type { InstructionEvent, SoundEvent, Opcode, Timbre } from '../types.ts';

/**
 * windy-lang opcode → sound parameters.
 *
 * docs/MAPPING.md is the source of truth; this file is that table
 * transcribed into code.
 */

interface SoundParams {
  frequency: number;
  duration: number;
  velocity: number;
  timbre: Timbre;
  /** End frequency, if this opcode glissandos. */
  endFrequency?: number;
}

const WIND_PARAMS: Record<
  Extract<
    Opcode,
    | 'MOVE_S'
    | 'MOVE_SW'
    | 'MOVE_W'
    | 'MOVE_NW'
    | 'MOVE_N'
    | 'MOVE_NE'
    | 'MOVE_E'
    | 'MOVE_SE'
  >,
  SoundParams
> = {
  MOVE_S: { frequency: 261.63, duration: 0.18, velocity: 0.5, timbre: 'sine' },
  MOVE_SW: { frequency: 293.66, duration: 0.18, velocity: 0.5, timbre: 'sine' },
  MOVE_W: { frequency: 329.63, duration: 0.18, velocity: 0.5, timbre: 'sine' },
  MOVE_NW: { frequency: 392.0, duration: 0.18, velocity: 0.5, timbre: 'sine' },
  MOVE_N: { frequency: 440.0, duration: 0.18, velocity: 0.5, timbre: 'sine' },
  MOVE_NE: { frequency: 523.25, duration: 0.18, velocity: 0.5, timbre: 'sine' },
  MOVE_E: { frequency: 587.33, duration: 0.18, velocity: 0.5, timbre: 'sine' },
  MOVE_SE: { frequency: 659.25, duration: 0.18, velocity: 0.5, timbre: 'sine' },
};

const DIGIT_FREQUENCIES = [
  523.25, // 0  C5
  587.33, // 1  D5
  659.25, // 2  E5
  783.99, // 3  G5
  880.0, // 4  A5
  1046.5, // 5  C6
  1174.66, // 6  D6
  1318.51, // 7  E6
  1567.98, // 8  G6
  1760.0, // 9  A6
];

const ARITH_PARAMS: Record<
  Extract<Opcode, 'ADD' | 'SUB' | 'MUL' | 'DIV' | 'MOD' | 'NOT' | 'GT'>,
  SoundParams
> = {
  ADD: { frequency: 392.0, duration: 0.08, velocity: 0.4, timbre: 'square' },
  SUB: { frequency: 349.23, duration: 0.08, velocity: 0.4, timbre: 'square' },
  MUL: { frequency: 523.25, duration: 0.08, velocity: 0.4, timbre: 'square' },
  DIV: { frequency: 466.16, duration: 0.08, velocity: 0.4, timbre: 'square' },
  MOD: { frequency: 311.13, duration: 0.08, velocity: 0.4, timbre: 'square' },
  NOT: { frequency: 880.0, duration: 0.06, velocity: 0.35, timbre: 'square' },
  GT: { frequency: 987.77, duration: 0.06, velocity: 0.35, timbre: 'square' },
};

const STACK_PARAMS: Record<Extract<Opcode, 'DUP' | 'DROP' | 'SWAP'>, SoundParams> = {
  DUP: { frequency: 698.46, duration: 0.06, velocity: 0.4, timbre: 'square' },
  DROP: { frequency: 220.0, duration: 0.06, velocity: 0.35, timbre: 'square' },
  SWAP: { frequency: 587.33, duration: 0.06, velocity: 0.4, timbre: 'square' },
};

const IO_PARAMS: Record<Extract<Opcode, 'PUT_NUM' | 'PUT_CHR' | 'GET_NUM' | 'GET_CHR'>, SoundParams> =
  {
    PUT_NUM: { frequency: 523.25, duration: 0.15, velocity: 0.45, timbre: 'sawtooth' },
    PUT_CHR: { frequency: 880.0, duration: 0.1, velocity: 0.4, timbre: 'sawtooth' },
    GET_NUM: { frequency: 130.81, duration: 0.2, velocity: 0.5, timbre: 'sawtooth' },
    GET_CHR: { frequency: 659.25, duration: 0.05, velocity: 0.4, timbre: 'sawtooth' },
  };

/**
 * Translate one instruction into one (or a few) SoundEvents.
 *
 * Some opcodes need two voices (glissando, dual-tone). v1 keeps it
 * simple: a single SoundEvent represents the glissando and the synth
 * handles it via the endFrequency hint.
 */
export function mapInstruction(event: InstructionEvent, when: number): SoundEvent[] {
  const { opcode, digit, ipId } = event;
  const voiceId = ipId;

  switch (opcode) {
    case 'NOP':
      // Silence — produces no event at all.
      return [];

    case 'HALT':
      return [
        {
          frequency: 130.81, // C3
          duration: 0.4,
          velocity: 0.6,
          timbre: 'sine',
          when,
          voiceId,
        },
      ];

    case 'TRAMPOLINE':
      return [{ frequency: 800, duration: 0.05, velocity: 0.3, timbre: 'square', when, voiceId }];

    case 'SPLIT':
      // Sine an octave up — signals the polyphony fork.
      return [{ frequency: 880, duration: 0.2, velocity: 0.5, timbre: 'sine', when, voiceId }];

    case 'MOVE_S':
    case 'MOVE_SW':
    case 'MOVE_W':
    case 'MOVE_NW':
    case 'MOVE_N':
    case 'MOVE_NE':
    case 'MOVE_E':
    case 'MOVE_SE':
      return [{ ...WIND_PARAMS[opcode], when, voiceId }];

    case 'TURBULENCE':
      return [{ frequency: 0, duration: 0.15, velocity: 0.4, timbre: 'noise', when, voiceId }];

    case 'GUST':
      // Glissando 200 → 800; synth handles the sweep via endFrequency.
      return [
        {
          frequency: 200,
          duration: 0.18,
          velocity: 0.5,
          timbre: 'sine',
          when,
          voiceId,
        },
      ];

    case 'CALM':
      // Glissando 800 → 200.
      return [
        {
          frequency: 800,
          duration: 0.18,
          velocity: 0.5,
          timbre: 'sine',
          when,
          voiceId,
        },
      ];

    case 'PUSH_DIGIT': {
      const idx = digit ?? 0;
      const freq = DIGIT_FREQUENCIES[idx] ?? DIGIT_FREQUENCIES[0]!;
      return [{ frequency: freq, duration: 0.12, velocity: 0.45, timbre: 'triangle', when, voiceId }];
    }

    case 'STR_MODE':
      // String-mode toggle — short pluck.
      return [{ frequency: 440, duration: 0.06, velocity: 0.3, timbre: 'sawtooth', when, voiceId }];

    case 'ADD':
    case 'SUB':
    case 'MUL':
    case 'DIV':
    case 'MOD':
    case 'NOT':
    case 'GT':
      return [{ ...ARITH_PARAMS[opcode], when, voiceId }];

    case 'DUP':
    case 'DROP':
    case 'SWAP':
      return [{ ...STACK_PARAMS[opcode], when, voiceId }];

    case 'IF_H':
      // dual-tone 200 + 400
      return [
        { frequency: 200, duration: 0.1, velocity: 0.35, timbre: 'square', when, voiceId },
        { frequency: 400, duration: 0.1, velocity: 0.35, timbre: 'square', when, voiceId },
      ];

    case 'IF_V':
      // dual-tone 300 + 600
      return [
        { frequency: 300, duration: 0.1, velocity: 0.35, timbre: 'square', when, voiceId },
        { frequency: 600, duration: 0.1, velocity: 0.35, timbre: 'square', when, voiceId },
      ];

    case 'PUT_NUM':
    case 'PUT_CHR':
    case 'GET_NUM':
    case 'GET_CHR':
      return [{ ...IO_PARAMS[opcode], when, voiceId }];

    case 'GRID_GET':
      return [{ frequency: 392, duration: 0.1, velocity: 0.4, timbre: 'sawtooth', when, voiceId }];

    case 'GRID_PUT':
      return [{ frequency: 523, duration: 0.1, velocity: 0.4, timbre: 'sawtooth', when, voiceId }];
  }
}

/**
 * Default per-opcode duration — referenced by the sequencer when it
 * decides the `when` of the next event.
 */
export function defaultDuration(opcode: Opcode): number {
  // NOP is a short rest, not zero — gives the melody room to breathe.
  if (opcode === 'NOP') return 0.1;
  if (opcode === 'HALT') return 0.4;

  // Winds 0.18, digits 0.12, arithmetic/stack 0.08, everything else ~0.1.
  const winds: Opcode[] = [
    'MOVE_S',
    'MOVE_SW',
    'MOVE_W',
    'MOVE_NW',
    'MOVE_N',
    'MOVE_NE',
    'MOVE_E',
    'MOVE_SE',
    'TURBULENCE',
    'GUST',
    'CALM',
  ];
  if (winds.includes(opcode)) return 0.18;
  if (opcode === 'PUSH_DIGIT') return 0.12;
  if (opcode === 'SPLIT') return 0.2;

  const short: Opcode[] = ['ADD', 'SUB', 'MUL', 'DIV', 'MOD', 'NOT', 'GT', 'DUP', 'DROP', 'SWAP'];
  if (short.includes(opcode)) return 0.08;

  if (opcode === 'TRAMPOLINE') return 0.06;

  return 0.1;
}

File: cli/src/mapping.rs
//! Opcode → SoundParams lookup.
//!
//! Direct port of `src/audio/mapping.ts` from the parent web app —
//! same frequencies, durations, velocities, timbres. The web version
//! is the canonical reference; the CLI tracks it on a per-release
//! basis (no shared spec file yet).

use windy::opcodes::Op;

use crate::trace::InstructionEvent;

#[derive(Clone, Copy, Debug)]
pub enum Timbre {
    Sine,
    Triangle,
    Square,
    Sawtooth,
    Noise,
}

#[derive(Clone, Copy, Debug)]
pub struct SoundEvent {
    pub when: f32,      // seconds from render start
    pub frequency: f32, // Hz (start freq for glissando)
    pub end_frequency: Option<f32>, // Hz (end freq for glissando)
    pub duration: f32,  // seconds
    pub velocity: f32,  // 0..1
    pub timbre: Timbre,
    /// Which IP this event belongs to. Held for future panning / channel
    /// routing — v0.1 renders mono so it's not read yet.
    #[allow(dead_code)]
    pub voice_id: u32,
}

/// PushDigit lands on one of these — mirrors the TS DIGIT_FREQUENCIES table.
const DIGIT_FREQ: [f32; 10] = [
    523.25,  // 0  C5
    587.33,  // 1  D5
    659.25,  // 2  E5
    783.99,  // 3  G5
    880.00,  // 4  A5
    1046.50, // 5  C6
    1174.66, // 6  D6
    1318.51, // 7  E6
    1567.98, // 8  G6
    1760.00, // 9  A6
];

/// Map a single VM event into zero or one rendered sound. `when` is
/// the seconds-offset from the start of the rendered audio.
pub fn map_instruction(event: &InstructionEvent, when: f32) -> Option<SoundEvent> {
    let voice_id = event.ip_id;

    // String mode: every character pushes its codepoint silently. The
    // web app v1 also stays silent here — sonification of in-string
    // text is out of scope.
    if event.strmode {
        return None;
    }

    let params = match event.opcode {
        Op::Nop | Op::Unknown => return None,

        Op::Halt => fixed(130.81, 0.4, 0.6, Timbre::Sine),
        Op::Trampoline => fixed(800.0, 0.05, 0.3, Timbre::Square),
        Op::Split => fixed(880.0, 0.2, 0.5, Timbre::Sine),
        Op::Turbulence => fixed(660.0, 0.08, 0.4, Timbre::Noise),

        Op::MoveS => fixed(261.63, 0.18, 0.5, Timbre::Sine),
        Op::MoveSw => fixed(293.66, 0.18, 0.5, Timbre::Sine),
        Op::MoveW => fixed(329.63, 0.18, 0.5, Timbre::Sine),
        Op::MoveNw => fixed(392.00, 0.18, 0.5, Timbre::Sine),
        Op::MoveN => fixed(440.00, 0.18, 0.5, Timbre::Sine),
        Op::MoveNe => fixed(523.25, 0.18, 0.5, Timbre::Sine),
        Op::MoveE => fixed(587.33, 0.18, 0.5, Timbre::Sine),
        Op::MoveSe => fixed(659.25, 0.18, 0.5, Timbre::Sine),

        Op::PushDigit => {
            let d = event.digit.unwrap_or(0).min(9) as usize;
            fixed(DIGIT_FREQ[d], 0.10, 0.45, Timbre::Triangle)
        }

        Op::StrMode => fixed(220.0, 0.04, 0.25, Timbre::Square),

        Op::Add => fixed(392.00, 0.08, 0.40, Timbre::Square),
        Op::Sub => fixed(349.23, 0.08, 0.40, Timbre::Square),
        Op::Mul => fixed(523.25, 0.08, 0.40, Timbre::Square),
        Op::Div => fixed(466.16, 0.08, 0.40, Timbre::Square),
        Op::Mod => fixed(311.13, 0.08, 0.40, Timbre::Square),
        Op::Not => fixed(880.00, 0.06, 0.35, Timbre::Square),
        Op::Gt => fixed(987.77, 0.06, 0.35, Timbre::Square),

        Op::Dup => fixed(698.46, 0.06, 0.40, Timbre::Square),
        Op::Drop => fixed(220.00, 0.06, 0.35, Timbre::Square),
        Op::Swap => fixed(587.33, 0.06, 0.40, Timbre::Square),

        Op::IfH => fixed(440.00, 0.08, 0.45, Timbre::Triangle),
        Op::IfV => fixed(660.00, 0.08, 0.45, Timbre::Triangle),

        Op::PutNum => fixed(523.25, 0.15, 0.45, Timbre::Sawtooth),
        Op::PutChr => fixed(880.00, 0.10, 0.40, Timbre::Sawtooth),
        Op::GetNum => fixed(130.81, 0.20, 0.50, Timbre::Sawtooth),
        Op::GetChr => fixed(659.25, 0.05, 0.40, Timbre::Sawtooth),

        Op::GridGet => fixed(196.00, 0.10, 0.45, Timbre::Triangle),
        Op::GridPut => fixed(247.00, 0.10, 0.45, Timbre::Triangle),

        // GUST/CALM are magic-frequency markers that the synth layer
        // turns into glissandos. Encode as start/end pair here.
        Op::Gust => gliss(200.0, 800.0, 0.25, 0.55, Timbre::Sine),
        Op::Calm => gliss(800.0, 200.0, 0.25, 0.55, Timbre::Sine),
    };

    Some(SoundEvent {
        when,
        frequency: params.frequency,
        end_frequency: params.end_frequency,
        duration: params.duration,
        velocity: params.velocity,
        timbre: params.timbre,
        voice_id,
    })
}

// Small intermediate builder so the match arms read clean.
struct Params {
    frequency: f32,
    end_frequency: Option<f32>,
    duration: f32,
    velocity: f32,
    timbre: Timbre,
}

fn fixed(frequency: f32, duration: f32, velocity: f32, timbre: Timbre) -> Params {
    Params {
        frequency,
        end_frequency: None,
        duration,
        velocity,
        timbre,
    }
}

fn gliss(from: f32, to: f32, duration: f32, velocity: f32, timbre: Timbre) -> Params {
    Params {
        frequency: from,
        end_frequency: Some(to),
        duration,
        velocity,
        timbre,
    }
}
